bindings/python/cpp/src/ghex.cpp

```diff
@@ -0,0 +1,130 @@
+    };
+    using pattern_type = typename std::result_of<decltype(make_pattern_wrapper)(context_type&, HaloGenerator&, std::vector<DomainRange>&)>::type;
+
+    m.def("make_pattern", make_pattern_wrapper);
+
+    // todo: move + cleanup
+    using arch = gridtools::ghex::cpu;
+    using domain_id_type = int;
+    using domain_descriptor_type = gridtools::ghex::structured::regular::domain_descriptor<domain_id_type, std::integral_constant<int, 3>>;
+    using array_type = std::array<int, 3>;
+    auto wrapper = [] (const domain_descriptor_type& dom, double* data, const array_type& offsets, const array_type& extents) {
+        return gridtools::ghex::wrap_field<arch, ::gridtools::layout_map<0,1,2>>(dom, data, offsets, extents);
+    };
+    using field_t = decltype(wrapper(std::declval<const domain_descriptor_type&>(), std::declval<double*>(), std::declval<const array_type&>(), std::declval<const array_type&>()));
+
+    py::class_<pattern_type>(m, "PatternContainer")
+        .def("__call__", &pattern_type::template operator()<field_t>);
+
+    using buffer_info_type = decltype(std::declval<pattern_type>()(std::declval<field_t&>()));
+    py::class_<buffer_info_type>(m, "BufferInfo");
+}
+
+
+template <typename GridType, typename Transport, typename HaloGenerator, typename Domain>
+void register_make_communication_object(pybind11::module& m) {
+    using grid_type = typename GridType::template type<Domain>;
+    using domain_id_type = typename Domain::domain_id_type;
+
+
+    using context_type = typename gridtools::ghex::tl::context_factory<Transport>::context_type;
+    using pattern_type = decltype(gridtools::ghex::make_pattern<GridType>(std::declval<context_type&>(), std::declval<HaloGenerator&>(), std::declval<std::vector<Domain>&>()));
+    using communicator_type = typename context_type::communicator_type;
+    using communication_object_type = gridtools::ghex::communication_object<communicator_type, grid_type, domain_id_type>;
+
+
+    // todo: just for getting something runnable
+    using arch = gridtools::ghex::cpu;
+    using domain_descriptor_type = gridtools::ghex::structured::regular::domain_descriptor<domain_id_type, std::integral_constant<int, 3>>;
+    using array_type = std::array<int, 3>;
+    auto wrapper = [] (const domain_descriptor_type& dom, double* data, const array_type& offsets, const array_type& extents) {
+        return gridtools::ghex::wrap_field<arch, ::gridtools::layout_map<0,1,2>>(dom, data, offsets, extents);
+    };
+    using field_t = decltype(wrapper(std::declval<const domain_descriptor_type&>(), std::declval<double*>(), std::declval<const array_type&>(), std::declval<const array_type&>()));
+
+    using buffer_info_type = decltype(std::declval<pattern_type>()(std::declval<field_t&>()));
+
+    py::class_<communication_object_type>(m, "CommunicationObject")
+        .def(py::init([] (communicator_type c) {
+            return gridtools::ghex::make_communication_object<pattern_type>(c); }))
```

Comment:
In principle, could a similar approach (make_communication_object... called within python init function) be used also before, for the pattern?